metagraph/src/graph/representation/hash/dbg_sshash.cpp

```diff
@@ -0,0 +1,494 @@
+#include "dbg_sshash.hpp"
+
+#include <sdsl/uint256_t.hpp>
+
+#include "common/seq_tools/reverse_complement.hpp"
+#include "common/threads/threading.hpp"
+#include "common/logger.hpp"
+#include "kmer/kmer_extractor.hpp"
+
+
+namespace mtg {
+namespace graph {
+
+const std::string DBGSSHash::alphabet_ = kmer::KmerExtractor2Bit().alphabet;
+
+using Kmer64 = uint64_t;
+using Kmer128 = __uint128_t;
+using Kmer256 = sdsl::uint256_t;
+
+template <typename KmerInt>
+class Dictionary : public IDictionary {
```

Comment:
Should we maybe have a base untemplated class in sshash fork instead, and inherit `sshash::dictionary` from it?

Reply:
Or, alternatively, use the approach similar to [dbg_hash_fast.cpp](https://github.com/ratschlab/metagraph/blob/master/metagraph/src/graph/representation/hash/dbg_hash_fast.cpp), that is structure `DBGSSHash` this way, rather than `sshash::dictionary`? Then it'd be more consistent with how other DBG implementations are done.

Reply:
Or maybe use
```cpp
std::variant<
    sshash::dictionary<Kmer64>,
    sshash::dictionary<Kmer128>,
    sshash::dictionary<Kmer256>>
```

with `std::visit`, to reduce overhead in polymorphysm of dictionary? It may simplify [this](https://github.com/ratschlab/metagraph/pull/481/files#diff-e90dbf65e72f4eaf563486b42faf095ec39582fe0b24acacea4b1bc3a7b6d79fR194-R199) and similar checks, and also remove the need of repeatedly declaring functions that just mirror base behavior, if I understand correctly.

Reply:
I'd prefer to have an untemplated parent class in sshash to handle this, then we can initialize the dictionary with the appropriate template in the constructer and load. For this to work consistently, however, we'd need to implement a sequential access method to avoid exposing `kmer_t`.

I don't really like the way it's done in DBGHashFast since there are a lot of methods that need to be implemented, so it produces a lot of extra code.

Reply:
What do you think about using `std::variant<sshash::dictionary<...>>` instead? Is there a reason not to do it? The problem with introducing untemplated parent class to sshash itself, from my perspective, would be that sshash itself currently doesn't seem to have a need for it, so it'd be an unnecessary complication of their codebase.

Reply:
ok, I implemented this